File: 494.Target Sum/solution.py
class Solution(object):
    def findTargetSumWays(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        dp = {}
        def back(index,total):
            if index == len(nums):
                if total==target:
                    return 1
                return 0
            
            if (index,total) in dp:
                return dp[(index,total)]
            
            dp[(index, total)] = back(index+1,total+nums[index])+back(index+1,total-nums[index])
            return dp[(index,total)]
        return back(0,0)

File: 494.Target Sum/solution.java
class Solution {
    public int helper(int ind, int[] arr, int sum, int target) {
        if(ind == arr.length) {

            if(sum == target) {
                return 1;
            }
            return 0;
        }
        int add = helper(ind + 1, arr, sum + arr[ind], target);
        int sub = helper(ind + 1, arr, sum - arr[ind], target);
        return add + sub;
    }
    public int findTargetSumWays(int[] nums, int target) {
        // int count  = 0;
        return helper(0, nums, 0, target);
        // return count;
    }
}